src/audit/unpinned_container_images.rs

```diff
@@ -0,0 +1,93 @@
+use anyhow::Result;
+
+use crate::{
+    finding::{Confidence, Finding, Severity, SymbolicLocation},
+    models::{Job, JobExt as _, Workflow},
+    state::AuditState,
+};
+
+use github_actions_models::workflow::job::Container;
+
+use super::{Audit, AuditLoadError, audit_meta};
+
+pub(crate) struct UnpinnedContainerImages;
+
+impl UnpinnedContainerImages {
+    fn container_image_ref_is_unpinned<'doc>(&self, image: &str) -> bool {
+        !image.contains(':')
+    }
+
+    fn container_image_ref_is_pinned_to_latest<'doc>(&self, image: &str) -> bool {
+        image.ends_with(":latest")
+    }
+
+    fn build_finding<'doc>(&self, image: &str, location: SymbolicLocation<'doc>, workflow: &'doc Workflow) -> Result<Finding<'doc>> {
+        let mut annotated_location = location;
+        if self.container_image_ref_is_unpinned(image) {
+            annotated_location = annotated_location.annotated("container image is unpinned");
+        }
+        if self.container_image_ref_is_pinned_to_latest(image) {
+            annotated_location = annotated_location.annotated("container image is pinned to latest");
+        }
+        Self::finding()
+            .severity(Severity::High)
+            .confidence(Confidence::High)
+            .add_location(annotated_location)
+            .build(workflow)
+    }
+}
+
+audit_meta!(
+    UnpinnedContainerImages,
+    "unpinned-container-images",
+    "unpinned container image references"
+);
+
+impl Audit for UnpinnedContainerImages {
+    fn new(_state: &AuditState<'_>) -> Result<Self, AuditLoadError> {
+        Ok(Self)
+    }
+
+    fn audit_workflow<'doc>(&self, workflow: &'doc Workflow) -> Result<Vec<Finding<'doc>>> {
```

Comment:
FYI: you can use `audit_normal_job` instead of `audit_workflow` to have `zizmor`'s machinery do the job loop + filter for you 🙂 -- see `bot_conditions.rs` for an example of that.

(That'll make the implementation here a bit shorter.)